### ISSUE TITLE
False positives for `ClassMustBeFinal` with cache

### ISSUE THREAD

**Issue by iquito** [vimeo/psalm#11299]
Since `ClassMustBeFinal` was introduced, I keep getting false positives for it when running Psalm with cache enabled, like in this blatant example:

```php
ERROR: ClassMustBeFinal - src/Controller/Reseller/PortingController.php:20:13 - Class Controller\Reseller\PortingController is never extended and is not part of the public API, and thus must be made final. (see https://psalm.dev/361)
final class PortingController extends AbstractResellerController
```

As you can see, the class already has the final keyword (even includes it in the snippet of Psalm itself), yet the error comes up. The same has happened many times for a class that is extended by other classes, but suddenly Psalm reports I should mark it as final because it is not extended (even though since running Psalm I did not change either the base class nor the extended classes).

This seems to be somehow related to the cache, because when I run Psalm without the cache / after clearing the cache, all these issues disappear. Somehow `ClassMustBeFinal` seems to incorrectly handle the cache or there is a weird interaction with it where some information is not used or lost - creating a test case for it is difficult though, as I don't see a clear origin of when and how such false positives come up.

**Comment by danog** [comment]
Probably something off in the population phase, generally I would recommend always running Psalm without cache enabled.

Will work more on cache in the coming weeks for a client, in the meantime I would 100% recommend using `--no-cache` (enabling JIT instead and running on UNIX systems to improve performance).

**Comment by orklah** [comment]
I also had some of those while debugging in Psalm itself. It seemed to consistantly happen after a crash in Psalm, as if the cache was being corrupted.

Maybe it can help with debugging...

**Comment by Tofandel** [comment]
Same issues with other rules

After running multiple times psalm I got

```Method App\Utils\Utils::relativeUrl is not static, but is called statically```

```php
    public static function relativeUrl(?string $absoluteUrl): ?string
    {
```
